openpype/pipeline/workfile/abstract_template_loader.py

```diff
@@ -0,0 +1,483 @@
+
+    def load_succeed(self, placeholder, container):
+        placeholder.parent_in_hierarchy(container)
+
+    def load_failed(self, placeholder, last_representation):
+        self.log.warning("Got error trying to load {}:{} with {}\n\n"
+                         "{}".format(last_representation['context']['asset'],
+                                     last_representation['context']['subset'],
+                                     placeholder.loader,
+                                     traceback.format_exc()))
+
+    def postload(self, placeholder):
+        placeholder.clean()
+
+    def update_missing_containers(self):
+        loaded_containers_ids = self.get_loaded_containers_by_id()
+        self.populate_template(ignored_ids=loaded_containers_ids)
+
+    def get_placeholders(self):
+        placeholder_class = self.placeholder_class
+        placeholders = map(placeholder_class, self.get_template_nodes())
+        valid_placeholders = filter(placeholder_class.is_valid, placeholders)
+        sorted_placeholders = sorted(valid_placeholders,
+                                     key=placeholder_class.order)
+        return sorted_placeholders
+
+    @abstractmethod
+    def get_loaded_containers_by_id(self):
+        """
+        Collect already loaded containers for updating scene
+        Return:
+            dict (string, node): A dictionnary id as key
+            and containers as value
+        """
+        pass
+
+    @abstractmethod
+    def import_template(self, template_path):
+        """
+        Import template in current host
+        Args:
+            template_path (str): fullpath to current task and
+                host's template file
+        Return:
+            None
+        """
+        pass
+
+    @abstractmethod
+    def get_template_nodes(self):
+        """
+        Returning a list of nodes acting as host placeholders for
+        templating. The data representation is by user.
+        AbstractLoadTemplate (and LoadTemplate) won't directly manipulate nodes
+        Args :
+            None
+        Returns:
+            list(AnyNode): Solved template path
+        """
+        pass
+
+
+@six.add_metaclass(ABCMeta)
+class AbstractPlaceholder:
+    """Abstraction of placeholders logic
+    Properties:
+        attributes: A list of mandatory attribute to decribe placeholder
+            and assets to load.
+        optional_attributes: A list of optional attribute to decribe
+            placeholder and assets to load
+        loader: Name of linked loader to use while loading assets
+        is_context: Is placeholder linked
+            to context asset (or to linked assets)
+    Methods:
+        is_repres_valid:
+        loader:
+        order:
+        is_valid:
+        get_data:
+        parent_in_hierachy:
+    """
+
+    attributes = {'builder_type', 'family', 'representation',
+                  'order', 'loader', 'loader_args'}
+    optional_attributes = {}
+
+    def __init__(self, node):
```

Comment:
I'm not sure what the `node` should represent, is it node in workfile? We should consider different variable name with some docstrings.